File: alpha/experiments/item-storage-protocol-standalone/src/exp/e6_dial.ts
// E6. The dial — assurance is a declared setting with a true, linear cost. We
// define audit tiers (monthly/weekly/daily/hourly, at k=5 or k=20), price each
// tier from E5's cost model (bytes read per audit) plus a fixed per-audit
// overhead, record the customer's chosen tier as a signed declaration, run one
// period per tier billed through the statement chain, and show a mid-period dial
// change pro-rates correctly. No judgment is encoded either way: your paranoia,
// your bill.

import type { World } from "../world.ts";
import { Checker, type ExperimentResult } from "../types.ts";
import { rentCents, auditCents, AUDIT_OVERHEAD_CENTS } from "../pricing.ts";

const PERIOD = 30;

type Tier = { name: string; k: number; auditsPerPeriod: number };

const TIERS: Tier[] = [
  { name: "monthly", k: 5, auditsPerPeriod: 1 },
  { name: "weekly", k: 5, auditsPerPeriod: 4 },
  { name: "daily", k: 20, auditsPerPeriod: 30 },
  { name: "hourly", k: 20, auditsPerPeriod: 720 },
];

export function run(world: World): ExperimentResult {
  const c = new Checker();
  const { customer, manifest } = world;

  // Representative bytes per audited item: the manifest average. SEAM: the demo
  // corpus is tiny, so audit cost is priced from k * avg-item-size (the E5 cost
  // model) rather than executed against it; the pricing is the point of E6.
  const avgItemBytes = Math.floor(manifest!.totalBytes / manifest!.leaves.length);

  const perAuditCents = (k: number) => auditCents(k * avgItemBytes, 1);
  const tierCost = (t: Tier) => t.auditsPerPeriod * perAuditCents(t.k);

  let day = world.clock.now(); // day 90 after E4
  const rows: (string | number)[][] = [];

  for (const tier of TIERS) {
    // Ada records her chosen tier as a signed declaration before the period runs.
    customer.ledger.append("dial-declaration", world.clock.iso(), {
      tier: tier.name, k: tier.k, auditsPerPeriod: tier.auditsPerPeriod,
    }, [customer.signer()]);

    const start = day;
    world.clock.advanceDays(PERIOD);
    day = world.clock.now();
    const byteDays = world.byteDays(start, day);
    const rc = rentCents(byteDays);
    const auditBytes = tier.auditsPerPeriod * tier.k * avgItemBytes;
    const ac = tierCost(tier);
    const stmt = world.commitStatement({
      periodStartDay: start, periodEndDay: day,
      openingRoot: manifest!.root, closingRoot: manifest!.root, byteDays,
      rentCents: rc, postageBytes: 0, postageCents: 0,
      auditCount: tier.auditsPerPeriod, auditBytes, auditCents: ac, auditTier: tier.name,
      graceCents: 0, feesCents: 0, totalCents: rc + ac,
    });
    rows.push([tier.name, tier.k, tier.auditsPerPeriod, ac, stmt.totalCents]);

    // The chosen tier appears in the statement.
    c.eq(`statement records the chosen tier (${tier.name})`, stmt.auditTier, tier.name);
  }

  // Billed audit cost is linear in audit count (same k => cost scales with count).
  // monthly vs weekly share k=5; weekly runs 4x the audits => 4x the audit cost.
  c.eq("weekly audit cost = 4 x monthly (same k, linear in count)",
    tierCost(TIERS[1]), 4 * tierCost(TIERS[0]));
  // daily vs hourly share k=20; hourly runs 24x the audits => 24x the cost.
  c.eq("hourly audit cost = 24 x daily (same k, linear in count)",
    tierCost(TIERS[3]), 24 * tierCost(TIERS[2]));

  // Changing the dial mid-period pro-rates correctly: weekly for the first half,
  // daily for the second half. Audit counts pro-rate by days.
  const weekly = TIERS[1], daily = TIERS[2];
  const auditsFor = (t: Tier, days: number) => Math.round((t.auditsPerPeriod * days) / PERIOD);
  const start = day;
  world.clock.advanceDays(15);
  const firstHalfAudits = auditsFor(weekly, 15); // 2
  world.clock.advanceDays(15);
  const secondHalfAudits = auditsFor(daily, 15); // 15
  day = world.clock.now();
  const firstCost = firstHalfAudits * perAuditCents(weekly.k);
  const secondCost = secondHalfAudits * perAuditCents(daily.k);
  const proratedAudits = firstHalfAudits + secondHalfAudits;
  const proratedCost = firstCost + secondCost;
  const byteDays = world.byteDays(start, day);
  const rc = rentCents(byteDays);
  const auditBytes = firstHalfAudits * weekly.k * avgItemBytes + secondHalfAudits * daily.k * avgItemBytes;
  const stmt = world.commitStatement({
    periodStartDay: start, periodEndDay: day,
    openingRoot: manifest!.root, closingRoot: manifest!.root, byteDays,
    rentCents: rc, postageBytes: 0, postageCents: 0,
    auditCount: proratedAudits, auditBytes, auditCents: proratedCost, auditTier: "weekly->daily",
    graceCents: 0, feesCents: 0, totalCents: rc + proratedCost,
  });
  c.eq("mid-period dial change bills the sum of the two pro-rated legs",
    stmt.auditCents, firstCost + secondCost);
  c.ok("pro-rated bill sits between a full-weekly and a full-daily period",
    proratedCost > tierCost(weekly) && proratedCost < tierCost(daily));
  rows.push(["weekly->daily", "5/20", proratedAudits, proratedCost, stmt.totalCents]);

  return {
    id: "E6",
    title: "The dial",
    plainSentence: "Declare your setting, pay its true cost, no judgment encoded either way.",
    assertions: c.results,
    tables: [
      {
        title: "Audit tiers billed through the statement (cents)",
        headers: ["tier", "k", "audits/period", "audit cost", "period total"],
        rows,
      },
    ],
    notes: [`Per-audit overhead: ${AUDIT_OVERHEAD_CENTS} cents; avg item ~${avgItemBytes} bytes.`],
  };
}

File: alpha/experiments/item-store/tests/e6_dial.rs
//! Phase 5 wiring test — ports E6 (the dial) from the oracle
//! (`src/exp/e6_dial.ts` + `src/pricing.ts`): audit assurance is a declared,
//! signed setting with a true, linear cost. Tiers are priced from the E5 cost
//! model (bytes read per audit) + a fixed per-audit overhead; the chosen tier is
//! a signed ledger declaration; audit cost is exactly linear in the audit count
//! (no discount/step at the tier edges); the chosen tier appears in the period
//! statement; and a mid-period dial change bills the sum of its two pro-rated
//! legs.
//!
//! RED→GREEN gate: exercises dial pricing + the signed declaration (into the
//! real ledger, re-verified) + the statement `audit_*` fields.

use std::collections::BTreeMap;

use item_store::crypto::derive_keypair;
use item_store::dial::{
    audits_for, declare_tier, per_audit_cents, tier_cost, DialDeclaration, Tier,
};
use item_store::identity::derive_id;
use item_store::ledger::{verify_entries, Ledger, Signer};
use item_store::statements::{build_statement, StatementBody, GENESIS_STATEMENT};

const AVG_ITEM_BYTES: u64 = 256;
const PERIOD: u64 = 30;

fn tiers() -> Vec<Tier> {
    vec![
        Tier::new("monthly", 5, 1),
        Tier::new("weekly", 5, 4),
        Tier::new("daily", 20, 30),
        Tier::new("hourly", 20, 720),
    ]
}

#[test]
fn audit_cost_is_exactly_linear_in_count_at_the_tier_edges() {
    let t = tiers();
    // weekly vs monthly share k=5; weekly runs 4x the audits => exactly 4x cost.
    assert_eq!(
        tier_cost(&t[1], AVG_ITEM_BYTES),
        4 * tier_cost(&t[0], AVG_ITEM_BYTES),
        "weekly = 4 x monthly (same k, linear in count)",
    );
    // daily vs hourly share k=20; hourly runs 24x the audits => exactly 24x cost.
    assert_eq!(
        tier_cost(&t[3], AVG_ITEM_BYTES),
        24 * tier_cost(&t[2], AVG_ITEM_BYTES),
        "hourly = 24 x daily (same k, linear in count)",
    );
    // No volume discount / step anywhere: N audits cost exactly N x one audit,
    // asserted at each tier's own count (the boundary sample points), not one
    // interior value. (Pass-3 mutation-resistance: kills a per-count discount.)
    for tier in &t {
        let one = per_audit_cents(tier.k, AVG_ITEM_BYTES);
        assert_eq!(
            tier_cost(tier, AVG_ITEM_BYTES),
            tier.audits_per_period * one,
            "{}: cost is exactly audits x per-audit (no discount)",
            tier.name,
        );
    }
}

#[test]
fn per_audit_cost_tracks_k_and_the_fixed_overhead() {
    // per-audit = floor(k*avg / 1000) + 2 overhead; avg=256:
    // k=5  -> floor(1280/1000)+2 = 3 ; k=20 -> floor(5120/1000)+2 = 7
    assert_eq!(per_audit_cents(5, AVG_ITEM_BYTES), 3);
    assert_eq!(per_audit_cents(20, AVG_ITEM_BYTES), 7);
    assert!(
        per_audit_cents(20, AVG_ITEM_BYTES) > per_audit_cents(5, AVG_ITEM_BYTES),
        "a larger k costs strictly more",
    );
}

#[test]
fn the_chosen_tier_is_a_signed_ledger_declaration() {
    let customer = derive_keypair("m", "customer");
    let id = derive_id(&customer.verifying_key());
    let mut ledger = Ledger::new(&id);
    let daily = Tier::new("daily", 20, 30);
    declare_tier(&mut ledger, &daily, "t0", Signer::new(&id, &customer));

    let entry = &ledger.entries()[0];
    assert_eq!(entry.kind, "dial-declaration");
    let decl: DialDeclaration =
        serde_json::from_value(entry.body.clone()).expect("declaration body round-trips");
    assert_eq!(decl.tier, "daily");
    assert_eq!(decl.k, 20);
    assert_eq!(decl.audits_per_period, 30);

    let ring: BTreeMap<String, _> = [(id.clone(), customer.verifying_key())]
        .into_iter()
        .collect();
    assert!(
        verify_entries(ledger.entries(), &ring).is_empty(),
        "the signed declaration verifies under the customer's key",
    );
}

#[test]
fn the_chosen_tier_appears_in_the_period_statement() {
    let daily = Tier::new("daily", 20, 30);
    let ac = tier_cost(&daily, AVG_ITEM_BYTES);
    let body = StatementBody {
        period: 0,
        period_start_day: 0,
        period_end_day: PERIOD,
        opening_root: "r".to_owned(),
        closing_root: "r".to_owned(),
        byte_days: 0,
        rent_cents: 0,
        postage_bytes: 0,
        postage_cents: 0,
        audit_count: daily.audits_per_period,
        audit_bytes: daily.audits_per_period * daily.k * AVG_ITEM_BYTES,
        audit_cents: ac,
        audit_tier: daily.name.clone(),
        grace_cents: 0,
        fees_cents: 0,
        total_cents: ac,
        prev_statement_hash: GENESIS_STATEMENT.to_owned(),
    };
    let stmt = build_statement(body);
    assert_eq!(stmt.body().audit_tier, "daily");
    assert_eq!(stmt.body().audit_cents, ac);
    assert_eq!(stmt.body().audit_count, 30);
}

#[test]
fn a_mid_period_dial_change_bills_the_sum_of_the_two_prorated_legs() {
    let weekly = Tier::new("weekly", 5, 4);
    let daily = Tier::new("daily", 20, 30);
    // First half weekly, second half daily; audit counts pro-rate by days.
    let first = audits_for(weekly.audits_per_period, 15, PERIOD); // round(4*15/30) = 2
    let second = audits_for(daily.audits_per_period, 15, PERIOD); // round(30*15/30) = 15
    assert_eq!(first, 2);
    assert_eq!(second, 15);

    let first_cost = first * per_audit_cents(weekly.k, AVG_ITEM_BYTES);
    let second_cost = second * per_audit_cents(daily.k, AVG_ITEM_BYTES);
    let prorated = first_cost + second_cost;

    // A full period at each tier, for the between-ness bound.
    let full_weekly = tier_cost(&weekly, AVG_ITEM_BYTES);
    let full_daily = tier_cost(&daily, AVG_ITEM_BYTES);
    assert!(
        prorated > full_weekly && prorated < full_daily,
        "pro-rated {prorated} sits between full-weekly {full_weekly} and full-daily {full_daily}",
    );
}
